### PR TITLE
research spike: when generating PoSt, reuse merkle trees generated during replication

### PR DESCRIPTION
**Pull request by laser**
## Why does this PR exist?

This changeset demonstrates a bare-bones merkle tree cache, and hopefully will give @schomatis some ideas for how rust-fil-proofs might integrate with the merkle_light changes he's made recently.

## Running the Code

To exercise this code, run:

```
RUST_LOG=INFO cargo test \
  --color=always \
  --release \
  --package filecoin-proofs \
  --lib api::tests::test_pip_lifecycle -- --nocapture --exact --ignored
```

### CODE REVIEW

File: filecoin-proofs/src/api/post.rs [hunk 1]
```diff
@@ -187,7 +191,48 @@ pub fn generate_post(
...
+        .map(|(id, replica)| {
+            // TODO: The cache-root should be provided by the caller, but for now we'll
+            // just create a directory in a known location
+            let mut cache = create_dumb_cache(id).unwrap();
+
+            let CachedTreeAbsPaths { top_half, leaves } = cache.paths(MerkleTreeCacheKey::CommR());
+
+            println!("checking merkle tree (leaves) cache-file: {:?}", leaves);
+
+            let tree: std::result::Result<_, failure::Error> = if Path::new(&leaves).exists() {
```
Comment (laser) [inline filecoin-proofs/src/api/post.rs]: @schomatis - This is where I attempt to read from the cache.

File: filecoin-proofs/src/api/post.rs [hunk 2]
```diff
@@ -187,7 +191,48 @@ pub fn generate_post(
...
+                    .create(true)
+                    .write(true)
+                    .truncate(false)
+                    .read(true)
+                    .open(top_half.clone())
+                    .unwrap();
+
+                top_half_cache_file.seek(SeekFrom::Start(0)).unwrap();
+
+                MerkleTree::new_with_files(
```
Comment (laser) [inline filecoin-proofs/src/api/post.rs]: @schomatis - This call is producing an `Err` in my branch. I suspect it has something to do with the permission bit used when I open the file(s).

File: storage-proofs/src/layered_drgporep.rs [hunk 3]
```diff
@@ -392,18 +442,49 @@ pub trait Layers {
...
+                        .create(true)
+                        .write(true)
+                        .truncate(true)
+                        .read(true)
+                        .open(leaves.clone())
+                        .unwrap();
+
+                    leaves_cache_file.seek(SeekFrom::Start(0)).unwrap();
+
+                    let mut leaves_store =
```
Comment (laser) [inline storage-proofs/src/layered_drgporep.rs]: @schomatis - This is where I populate the cache.

**Comment by schomatis** [comment]
@laser Looking good, taking ownership of this to study it in detail and make some changes, if there's nothing urgent don't push anymore to this PR until I ping you please.

**Comment by schomatis** [comment]
@laser I've updated the code to resolve the top-half build workflow, feel free to continue working on this PR.

**Comment by schomatis** [comment]
Test currently failing with:

```
creating (truncating) merkle tree cache-file: "/tmp/merkle-cache-SectorId(42)/v0-layer0-leaves"
creating (truncating) merkle tree cache-file: "/tmp/merkle-cache-SectorId(42)/v0-layer0-top"
creating (truncating) merkle tree cache-file: "/tmp/merkle-cache-SectorId(42)/v0-layer1-leaves"
creating (truncating) merkle tree cache-file: "/tmp/merkle-cache-SectorId(42)/v0-layer1-top"
creating (truncating) merkle tree cache-file: "/tmp/merkle-cache-SectorId(42)/v0-layer2-leaves"
creating (truncating) merkle tree cache-file: "/tmp/merkle-cache-SectorId(42)/v0-layer2-top"
creating (truncating) merkle tree cache-file: "/tmp/merkle-cache-SectorId(42)/v0-layer3-leaves"
creating (truncating) merkle tree cache-file: "/tmp/merkle-cache-SectorId(42)/v0-layer3-top"
creating (truncating) merkle tree cache-file: "/tmp/merkle-cache-SectorId(42)/v0-commr-leaves"
creating (truncating) merkle tree cache-file: "/tmp/merkle-cache-SectorId(42)/v0-commr-top"
test api::tests::test_pip_lifecycle ... test api::tests::test_pip_lifecycle has been running for over 60 seconds
checking merkle tree (leaves) cache-file: "/tmp/merkle-cache-SectorId(42)/v0-commr-leaves"
merkle tree (leaves) cache-file existed: "/tmp/merkle-cache-SectorId(42)/v0-commr-leaves"
loading merkle tree cache-file: "/tmp/merkle-cache-SectorId(42)/v0-commr-leaves"
loading merkle tree cache-file: "/tmp/merkle-cache-SectorId(42)/v0-commr-top"
Error: InvalidCommitment
thread 'api::tests::test_pip_lifecycle' panicked at 'assertion failed: `(left == right)`
  left: `1`,
 right: `0`: the test returned a termination value with a non-zero status code (1) which indicates a failure', src/libtest/lib.rs:336:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
test api::tests::test_pip_lifecycle ... FAILED

failures:

failures:
    api::tests::test_pip_lifecycle

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 19 filtered out

error: test failed, to rerun pass '-p filecoin-proofs --lib'
```

@laser Could you help me figure out this error? (The problem is most probably some bug in the MT code but I need some more information to be able to trace it.)

**Comment by schomatis** [comment]
Traced the error to:

https://github.com/filecoin-project/rust-fil-proofs/blob/f73147fb99e444f9649984865f049e12550bf541/storage-proofs/src/rational_post.rs#L136-L138

So the CommR tree is probably not being correctly loaded from the cache.

**Comment by schomatis** [comment]
@laser Cache test is working now. (Other tests may be failing as I drastically reduced many parameters to lower constraints and speed up parameter generation.)

```
running 1 test
creating (truncating) merkle tree cache-file: "/tmp/merkle-cache-sector-id-42/v0-layer0-leaves"
creating (truncating) merkle tree cache-file: "/tmp/merkle-cache-sector-id-42/v0-layer0-top"
creating (truncating) merkle tree cache-file: "/tmp/merkle-cache-sector-id-42/v0-layer1-leaves"
creating (truncating) merkle tree cache-file: "/tmp/merkle-cache-sector-id-42/v0-layer1-top"
creating (truncating) merkle tree cache-file: "/tmp/merkle-cache-sector-id-42/v0-layer2-leaves"
creating (truncating) merkle tree cache-file: "/tmp/merkle-cache-sector-id-42/v0-layer2-top"
creating (truncating) merkle tree cache-file: "/tmp/merkle-cache-sector-id-42/v0-layer3-leaves"
creating (truncating) merkle tree cache-file: "/tmp/merkle-cache-sector-id-42/v0-layer3-top"
creating (truncating) merkle tree cache-file: "/tmp/merkle-cache-sector-id-42/v0-commr-leaves"
creating (truncating) merkle tree cache-file: "/tmp/merkle-cache-sector-id-42/v0-commr-top"
checking merkle tree (leaves) cache-file: "/tmp/merkle-cache-sector-id-42/v0-commr-leaves"
merkle tree (leaves) cache-file existed: "/tmp/merkle-cache-sector-id-42/v0-commr-leaves"
loading merkle tree cache-file: "/tmp/merkle-cache-sector-id-42/v0-commr-leaves"
loading merkle tree cache-file: "/tmp/merkle-cache-sector-id-42/v0-commr-top"
test api::tests::test_pip_lifecycle ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 19 filtered out
```

**Comment by laser** [comment]
@dignifiedquire - May I close this PR? It has served its purpose.

**Comment by dignifiedquire** [comment]
close all the PRs

**Pull request closed without merging**